internal/alertmanager/currentalert/process.go

```diff
@@ -238,8 +238,61 @@ func addPvc(log logr.Logger, alertLabels model.LabelSet, alertAnnotations model.
 	return nil
 }
 
-func resizePvc(labels model.LabelSet, annotiations model.LabelSet, client client.Client) error {
-	//TODO
+func resizePvc(log logr.Logger, labels model.LabelSet, annotiations model.LabelSet, client client.Client) error {
+	cr, err := k8sutil.GetCr(string(labels["kafka_cr"]), string(labels["namespace"]), client)
+	if err != nil {
+		return err
+	}
+
+	pvc, err := getPvc(string(labels["persistentvolumeclaim"]), string(labels["namespace"]), client)
+	if err != nil {
+		return err
+	}
+
+	for i, broker := range cr.Spec.Brokers {
+		if strconv.Itoa(int(broker.Id)) == pvc.Labels["brokerId"] {
+			var storageConfigs []v1beta1.StorageConfig
+
+			if broker.BrokerConfig == nil {
+				storageConfigs = cr.Spec.BrokerConfigGroups[cr.Spec.Brokers[i].BrokerConfigGroup].StorageConfigs
+			} else {
+				storageConfigs = broker.BrokerConfig.StorageConfigs
+			}
+
+			for _, c := range storageConfigs {
+				if c.MountPath == pvc.Annotations["mountPath"] {
+					size := *c.PvcSpec.Resources.Requests.Storage()
+
+					incrementBy, err := resource.ParseQuantity(string(annotiations["incrementBy"]))
+					if err != nil {
+						return err
+					}
+
+					size.Add(incrementBy)
+
+					c.PvcSpec.Resources.Requests = corev1.ResourceList{
+						"storage": size,
+					}
+
+					if broker.BrokerConfig == nil {
+						broker.BrokerConfig = &v1beta1.BrokerConfig{
+							StorageConfigs: []v1beta1.StorageConfig{c},
```

Comment:
Let's assume that the storage configuration for the broker comes from a configuration group (thus `broker.BrokerConfig` is nil) and there are multiple log dirs (multiple PVCs) configured in the configuration group.
 
This means that we are going to have a slice of storage configs in the configuration group and a slice of storage configs with one element at the broker level. Since the two configs are merged using `WithAppendSlice` (https://github.com/banzaicloud/kafka-operator/blob/master/pkg/util/util.go#L193) this will lead to incorrect storage configuration.

```
...
spec:
  brokerConfigGroups:
    default:
      storageConfigs:
        - mountPath: "/kafka-logs-1"
          pvcSpec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 10Gi
        - mountPath: "/kafka-logs-2"
          pvcSpec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 10Gi
  brokers:
    - id: 0
      brokerConfigGroup: "default"
    - id: 1
      brokerConfigGroup: "default"
    - id: 2
      brokerConfigGroup: "default"
```

When an alert is triggered for let 's say the PVC that holds `/kafka-logs-2` this results
```
spec:
  brokerConfigGroups:
    default:
      storageConfigs:
        - mountPath: "/kafka-logs-1"
          pvcSpec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 10Gi
        - mountPath: "/kafka-logs-2"
          pvcSpec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 10Gi
  brokers:
    - id: 0
      brokerConfigGroup: "default"
      brokerConfig:
        storageConfigs:
          - mountPath: "/kafka-logs-2"
            pvcSpec:
              accessModes:
                - ReadWriteOnce
               resources:
                 requests:
                    storage: 25Gi
    - id: 1
      brokerConfigGroup: "default"
      brokerConfig:
        storageConfigs:
          - mountPath: "/kafka-logs-2"
            pvcSpec:
              accessModes:
                - ReadWriteOnce
               resources:
                 requests:
                    storage: 25Gi
    - id: 2
      brokerConfigGroup: "default"  
      brokerConfig:
        storageConfigs:
          - mountPath: "/kafka-logs-2"
            pvcSpec:
              accessModes:
                - ReadWriteOnce
               resources:
                 requests:
                    storage: 25Gi 
```

The `util.GetBrokerConfig` method will return a storage configuration slice that has `/kafka-logs-2` twice here https://github.com/banzaicloud/kafka-operator/blob/master/pkg/resources/kafka/kafka.go#L221 . This leads to non-deterministic behaviour.  In this case we should copy the entire storage configuration slice from the config group  and update the appropriate item in the slice.

Have you tested this scenario?

Reply:
Yes I have, there was discussion on this topic earlier on in this PR.

I updated the `utils.GetBrokerConfig` to dedup the storageConfigs based on the `mountPath` https://github.com/banzaicloud/kafka-operator/blob/101bc86d972cd59983cce899a8bcee81b3e73f17/pkg/util/util.go#L198

The broker specific configs take precedence over the `brokerConfigGroups` so in your example you would only end up reconciling one`/kafka-logs-2`. 

when we do: `err := mergo.Merge(bConfig, clusterSpec.BrokerConfigGroups[broker.BrokerConfigGroup], mergo.WithAppendSlice)` the bConfig is already inflated via a `deepCopy` if it is there. This will be in the first part of the new slice that `mergo` creates. 

I created a unit test to specific test this dedup/merge logic here: https://github.com/banzaicloud/kafka-operator/blob/101bc86d972cd59983cce899a8bcee81b3e73f17/pkg/util/util_test.go#L94

which covers the case of the path being defined in both the `brokerConfigGroup` and the `brokerConfig`. 

I also have this version of the operator deployed to a cluster managing 26 kafka clusters all with resized enabled so far 16 of them have had at least one volume resized. No issues as of yet (~1 week of running it).